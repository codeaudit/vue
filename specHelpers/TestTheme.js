import {setOpacity, toRgbaString} from '@andrew-codes/color-functions';

export default {
    _name: 'Test Theme',
    altColor: '#ffa500',
    basicColor: '#00ffff',
    basicFontFamily: '\'Proxima Nova\', \'Lucida Sans Unicode\', \'Lucida Grande\', sans-serif',
    baseIconSize: 24,
    boldFont: 600,
    borderPrimaryColor: '#000',
    darkInverseColor: '#fff',
    disabledPrimaryColor: '#808080',
    errorPrimaryColor: '#ffc0cb',
    errorSecondaryColor: '#ffc0cd',
    fieldBorderColor: '#808080',
    focusedPrimaryColor: '#00ffff',
    focusedSecondaryColor: toRgbaString(setOpacity('#00ffff', 0.5)),
    gutter: 8,
    importantColor: '#f99643',
    largeFontSize: 22,
    largeGutter: 12,
    largeLineHeight: 2.285,
    lightInverseColor: '#030303',
    mediumFontSize: 16,
    normalBackground: '#fff',
    normalLineHeight: 1.285,
    normalRadius: 3,
    requiredPrimaryColor: '#ffc0cb',
    pendingPrimaryColor: '#ffff00',
    smallFontSize: 14,
    smallGutter: 6,
    textPrimaryColor: '#030303',
    textSecondaryColor: '#030303',
    textDisabledColor: '#808080',
    xLargeFontSize: 24,
    xLargeGutter: 16,
    xMediumFontSize: 18,
    xSmallFontSize: 8,
    xxSmallGutter: 3
};
