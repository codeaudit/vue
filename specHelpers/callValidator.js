import ReactPropTypesArgs from './reactPropTypeArgs';

const callValidator = (validator, element, propName = '', componentName = '') => validator(element.props, propName, componentName, ...ReactPropTypesArgs);

export const validatorPasses = (validator, element, propName) => callValidator(validator, element, propName) === null;
export const validatorFails = (validator, element, propName) => callValidator(validator, element, propName) instanceof Error;
