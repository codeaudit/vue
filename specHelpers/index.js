import React from 'react';
import ThemeProvider, {getTheme} from './../src/Theme';

export const applyTheme = (ComponentUnderTest, theme = getTheme()) => <ThemeProvider theme={theme}>{ComponentUnderTest}</ThemeProvider>;
