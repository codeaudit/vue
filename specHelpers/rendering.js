import toJson from 'enzyme-to-json';
import {mount, shallow} from 'enzyme';
import React from 'react';
import theme from './TestTheme';

export const getRender = (Component) => (props = {}, context = {theme}) => shallow(<Component {...props} />, {context});
export const getMount = (Component) => (props = {}, context = {theme}) => mount(<Component {...props} />, {context});
export const snapshot = (renderedComponent) => toJson(renderedComponent);
