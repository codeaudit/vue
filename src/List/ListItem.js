import EventListener from 'react-event-listener';
import React, {Component, PropTypes} from 'react';
import Radium from './../utilities/Radium';
import transparent from './../utilities/Transparent';
import {createEventHandler} from './../utilities/component';

const getStyles = (props, theme) => {
    const {
        highlighted,
    } = props;
    const {
        smallGutter,
        largeGutter,
    } = theme;
    const hoveredStyles = highlighted
        ? {
            backgroundColor: props.highlightBackgroundColor,
            color: props.highlightColor,
        } : {
            backgroundColor: transparent,
            color: 'initial',
        };

    return {
        listItem: {
            cursor: 'pointer',
            padding: `${smallGutter}px ${largeGutter}px`,
            ...hoveredStyles,
        },
    };
};

const ListItem = (props, context) => {
    const {
        children,
        itemOid,
        onKeyUp,
        onMouseEnter,
    } = props;
    const handleMouseEnter = createEventHandler(onMouseEnter, itemOid);
    const handleKeyUp = createEventHandler(onKeyUp, itemOid);
    const styles = getStyles(props, context.theme);

    return (
        <div
            style={styles.listItem}
            onMouseEnter={handleMouseEnter}
            onKeyUp={handleKeyUp}
        >
            <EventListener
                target="window"
                onKeyUp={handleKeyUp}
            />
            {children}
        </div>
    );
};
ListItem.propTypes = {
    /**
     * Content to render within the list item
     */
    children: PropTypes.node.isRequired,
    /**
     * Color of the background when in a highlighted state
     */
    highlightBackgroundColor: PropTypes.string,
    /**
     * Color of the text when in a highlighted state
     */
    highlightColor: PropTypes.string,
    /**
     * When true, indicates the component is in a highlighted state
     */
    highlighted: PropTypes.bool,
    /**
     * Event handler; fired once the item is highlighted AND a key is pressed
     */
    onKeyUp: PropTypes.func,
    /**
     * Event handler; fired once mouse enters the component
     */
    onMouseEnter: PropTypes.func,
};
ListItem.defaultProps = {
    highlighted: false,
    onClick: () => {
    },
    onMouseEnter: () => {
    },
    onKeyUp: () => {
    },
};
ListItem.contextTypes = {
    theme: PropTypes.object.isRequired,
};
ListItem.displayName = 'ListItem';

export default Radium(ListItem);
