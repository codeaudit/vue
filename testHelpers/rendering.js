import toJson from 'enzyme-to-json';
import React, {PropTypes} from 'react';
import {mount, render, shallow} from 'enzyme';
import theme from './../packages/themes/src/V1Theme';

const ctx = {
    theme,
};

export const getShallow = (Component) => (props = {}, context = {}) => shallow(<Component {...props} />, {
    childContextTypes: {
        theme: PropTypes.object,
    },
    context: {
        ...ctx,
        ...context,
    },
});

export const getRender = (Component) => (props = {}, context = {}) => render(<Component {...props} />, {
    childContextTypes: {
        theme: PropTypes.object,
    },
    context: {
        ...ctx,
        ...context,
    },
});

export const getMount = (Component) => (props = {}, options = {}, context = {}) => mount(<Component {...props} />, {
    childContextTypes: {
        theme: PropTypes.object,
    },
    context: {
        ...ctx,
        ...context,
    },
    ...options,
});

const replaceUIKey = (componentName) => {
    let index = 0;
    return (json) => {
        if (!json.props || !json.props.uiKey) {
            return json;
        }
        const uiKey = `${componentName}-${index}`;
        index += 1;
        return Object.assign({}, json, {
            uiKey,
            uiPath: [uiKey],
        });
    };
};
const replaceUiKeyRecursively = (json) => {
    if (!json.type) {
        return json;
    }
    const replaceUIKeyInParent = replaceUIKey(json.type);
    const output = replaceUIKeyInParent(json);
    if (!output.children) {
        return json;
    }
    output.children = json.children.map((child, index) => replaceUiKeyRecursively(child));
    return output;
};

export const snapshot = (renderedComponent) => {
    return toJson(renderedComponent);
};

export const reset = (component) => () => {
    if (component) {
        component.unmount();
    }
    document.body.innerHTML = '';
};
