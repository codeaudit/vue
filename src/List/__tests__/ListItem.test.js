import ListItem from '../ListItem';
import {getShallow, snapshot} from './../../../specHelpers/rendering';

const shallowRenderListItem = getShallow(ListItem);
const evt = {test: true};

test('ListItem can render as not highlighted', () => {
    const component = shallowRenderListItem({
        children: 'List Item',
        highlightBackgroundColor: 'blue',
        highlightColor: 'pink',
        oid: 'ListItem',
    });
    expect(snapshot(component)).toMatchSnapshot();
});

test('ListItem can render as being highlighted', () => {
    const component = shallowRenderListItem({
        children: 'List Item',
        highlightBackgroundColor: 'blue',
        highlightColor: 'pink',
        highlighted: true,
        oid: 'ListItem',
    });
    expect(snapshot(component)).toMatchSnapshot();
});

test('ListItem responds to hover event', () => {
    const onMouseEnter = jest.fn();
    const component = shallowRenderListItem({
        children: 'List Item',
        oid: 'ListItem',
        onMouseEnter,
    });
    simulateHover(component, evt);
    expect(onMouseEnter).toHaveBeenCalledTimes(1);
    expect(onMouseEnter).toHaveBeenCalledWith(evt, 'ListItem');
});

function simulateHover(wrapper, evt = {}) {
    wrapper.simulate('mouseenter', evt);
}
