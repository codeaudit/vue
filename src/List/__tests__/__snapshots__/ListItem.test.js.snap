exports[`test ListItem can render as being highlighted 1`] = `
<div
  data-radium={true}
  onMouseEnter={[Function]}
  style={
    Object {
      "backgroundColor": "blue",
      "color": "pink",
      "cursor": "pointer",
      "padding": "6px 12px",
    }
  }>
  List Item
</div>
`;

exports[`test ListItem can render as not highlighted 1`] = `
<div
  data-radium={true}
  onMouseEnter={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0)",
      "color": "initial",
      "cursor": "pointer",
      "padding": "6px 12px",
    }
  }>
  List Item
</div>
`;
